editor-mode: prompt the user about overwriting file changes.

pdelfino opened the pull request:
Hopefully, this solves the feature request #1807 which makes the content on the [new article](https://github.com/atlas-engineer/nyxt-site/pull/14)  (private repo) ready to be published.

I removed the `(if-exists :error))` and `:if-exists if-exists`. Is it really necessary? Should I re-introduce it?

pdelfino commented:
@jmercouris, I have inserted the predicate suggested by you earlier today. 

The code works on the following scenarios:

1 - the user can make changes on a file that already exists. If they wanna save the changes, the prompt buffer is going to ask for a confirmation. if the answer is "yes", the changes will be overwritten;
2 - the user can make changes on a file that already exists. If they *did it accidentaly* and do **not** wanna save them, the prompt buffer is going to ask for a confirmation. if the answer is "no", the changes will **not** be overwritten;
3 - the user can open and add content on a file that **does not** exist yet. If they wanna save the new file with the content added, they just need to execute `editor-write-file` - no confirmation will be asked.

I think it is good to go :)

Ambrevar commented on source/editor-mode.lisp:
```diff
@@ -80,8 +80,13 @@ get/set-content (which is necessary for operation)."
-  (write-file-with-editor buffer :if-exists if-exists)
-  (echo "File ~a written to storage." (file buffer)))
+  (if (uiop:file-exists-p (file buffer))
+      (if-confirm ("Are you sure to overwrite file?")
```
Maybe just "Overwrite?".
What about showing the file name, for instance with `"Overwrite ~s?" (file buffer)`?

pdelfino replied:
I was just trying to mimic what already exists in the source code, especially the file `buffer.lisp` :

`(if-confirm ("Are you sure to delete ~a buffer~p?" count count)`

But I can change it to your suggestion. It would be more succint, indeed.

Ambrevar commented on source/editor-mode.lisp:
```diff
@@ -80,8 +80,13 @@ get/set-content (which is necessary for operation)."
+  (if (uiop:file-exists-p (file buffer))
+      (if-confirm ("Are you sure to overwrite file?")
+        (progn (write-file-with-editor buffer :if-exists :overwrite)
+               (echo "File ~a written to storage." (file buffer)))
```
Use `~s` instead to automatically surround the file with double quotes.  This is useful when the file has spaces.

Ambrevar commented on source/editor-mode.lisp:
```diff
@@ -80,8 +80,13 @@ get/set-content (which is necessary for operation)."
+      (if-confirm ("Are you sure to overwrite file?")
+        (progn (write-file-with-editor buffer :if-exists :overwrite)
+               (echo "File ~a written to storage." (file buffer)))
+        (echo "File ~a not written to storage." (file buffer)))
```
"written to storage" is not very idiomatic in my experience.  Simply `"File ~s saved."`?

pdelfino commented:
Just updated the PR with @Ambrevar's suggestions. I also changed an "unrelated" file (`buffer.lisp`) to make the format of the questions more consistent across the code base.

jmercouris commented on source/editor-mode.lisp:
```diff
@@ -80,8 +80,13 @@ get/set-content (which is necessary for operation)."
+        (progn (write-file-with-editor buffer :if-exists :overwrite)
+               (echo "File ~s saved." (file buffer)))
+        (echo "File ~s not saved." (file buffer)))
+      (progn  (write-file-with-editor buffer :if-exists if-exists)
```
It could be `:if-exists :error` instead, but that is purely a stylistic choice.
